File: src/main/java/de/fabilucius/advancedperks/configuration/types/AbstractMessageConfiguration.java
package de.fabilucius.advancedperks.configuration.types;

import de.fabilucius.advancedperks.configuration.Configuration;
import de.fabilucius.advancedperks.configuration.replace.ReplaceOptions;
import org.bukkit.ChatColor;

import java.util.List;

public abstract class AbstractMessageConfiguration extends Configuration {

    public String getComputedString(String key, ReplaceOptions... replaceOptions) {
        String message = this.getString(key, "message for key %s wasn't found".formatted(key));
        for (ReplaceOptions replaceOption : replaceOptions) {
            message = message.replaceAll(replaceOption.replaceKey(), replaceOption.replaceData());
        }
        message = message.replace("<prefix>", this.getPrefix());
        return ChatColor.translateAlternateColorCodes('&', message);
    }

    public List<String> getComputedStringList(String key, ReplaceOptions... replaceOptions) {
        List<String> messages = this.getStringList(key);
        if (messages.isEmpty()) {
            messages.add("message list for key %s wasn't found".formatted(key));
            return messages;
        }
        return messages.stream().map(message -> {
            for (ReplaceOptions replaceOption : replaceOptions) {
                message = message.replace(replaceOption.replaceKey(), replaceOption.replaceData());
            }
            message = message.replace("<prefix>", this.getPrefix());
            return ChatColor.translateAlternateColorCodes('&', message);
        }).toList();
    }

    public abstract String getPrefix();
}

File: src/main/kotlin/de/fabilucius/advancedperks/core/configuration/AbstractMessageConfiguration.kt
package de.fabilucius.advancedperks.core.configuration

import de.fabilucius.advancedperks.core.configuration.replace.ReplaceOptions
import de.fabilucius.advancedperks.core.logging.APLogger
import org.bukkit.ChatColor
import java.io.File

abstract class AbstractMessageConfiguration(configDir: File, logger: APLogger) : AbstractConfiguration(configDir, logger) {

    fun getMessageList(key: String, vararg replaceOptions: ReplaceOptions): List<String> =
        getStringList(key).takeIf { it.isNotEmpty() }?.let {
            it.map { message -> computeMessage(message, replaceOptions) }
        } ?: listOf("Message list for key $key not found.")

    fun getMessage(key: String, vararg replaceOptions: ReplaceOptions): String =
        getString(key)?.let {
            computeMessage(it, replaceOptions)
        } ?: "Message for key $key not found."

    private fun computeMessage(messageToCompute: String, replaceOptions: Array<out ReplaceOptions>): String =
        messageToCompute.let { message ->
            replaceOptions.fold(message) { acc, option ->
                acc.replace(option.replaceKey, option.replaceData)
            }.replace("<prefix>", getPrefix())
        }.let { ChatColor.translateAlternateColorCodes('&', it) }

    abstract fun getPrefix(): String

}